storybook and testing for plugins

illume opened the pull request:
# Allows using storybook and testing with plugins

- [x] Compiles the source from headlamp into /lib rather than just using types.
- [x] Can make .stories.tsx files inside plugins, and have them run with `npm run storybook` 
- [x] Documentation on how to write stories, with an example.
- [x] `npm run update-dependencies` script to update the headlamp-plugin/package.json dependencies to match frontend/package.json
- [x] `npm run check-dependencies` script to ensure dependencies are synced between headlamp-plugin/ and frontend/
- [x] storyshot snapshot testing works with `npm run test` inside the plugin
- [x] have an example github workflow for running tests
  - [x] use this for our testing our plugins
    - [x] will have to support running a directory of multiple plugins
  - [x] maybe can reuse some code from our test scripts, and move that into headlamp-plugin.
  - [x] have an example for folders of plugins, not just for one plugin
- [x] Add a `npm run storybook-build` to build storybook to static html.
  - [x] this provides some test coverage in a way, and is useful to check the separation of concerns. Because it's static html/js/css files it should work without other dependencies.
  - [x] It also can be useful to publish to a static website on PRs, for review.
- [x] make all the testing commands work on folders of packages (test, lint, format, storybook-build, tsc)
- [x] fix tsc command, so there's a `headlamp-plugin tsc` also so it works on folders of packages.
- [x] add a `headlamp-plugin format --check` to be used in the github workflow

This is also used to add github workflow to the plugins repo here: https://github.com/headlamp-k8s/plugins/pull/5

---

Task list for `npm run test`.

- [x] Separate PR for PluginLib.Cr change. Because that's required, and they use headlamp `main`. So that they can keep using latest main headlamp, and also use the pre-release @kinvolk/headlamp-plugin package.
- [x] `npm run test`
  - [x] encapsulate `npm run test` in headlamp-plugin
  - [x] upgrade should handle requirements for test.
    - [x] add a scripts `"test": "headlamp-plugin test"`
  - [x] copy in storyshot tests to pod-counter (storybook.test.tsx, setupTests.ts)
  - [x] see if we can incapsulate (storybook.test.tsx, setupTests.ts) inside headlamp-plugin
- [x] Get `import CRD, { KubeCRD, makeCustomResourceClass } from '@kinvolk/headlamp-plugin/lib/k8s/crd';` working. Use upgrade command to update import paths to new import path.
- [x] put frontend `lib/index.ts` back into headlamp plugin `src/index.ts`? I think it's better in src/index.ts because it's related to headlamp-plugin only. After the other restructuring, it now seems to make more sense in src/index... but would require retesting.
- [x] try removing Utils.ts CommonComponents.ts etc
  - [x] No, keep it there. Because it shows which ones are available in lib/.
  - [x] maybe move them into frontend/src/lib/ ?
- [x] have a --check command for update-dependencies.js and to CI. To make sure package dependencies are up to date. So it tells someone when frontend/package.json dependencies are changed that they need to also sync with headlamp-plugins (by using the update-dependencies script).
  - [x] Add yargs command line
- [x] include jest transformIgnorePatterns (maybe with a command line script inside headlamp-plugin). See pod-counter.

---



It completes these issues:
- setup testing https://github.com/kinvolk/headlamp/issues/569
- ships source in headlamp-plugin for testing #778


![headlamp-plugin-storybook](https://user-images.githubusercontent.com/9541/201612906-c189153c-5c55-4d0d-8bb4-c5dec743b94d.png)

## How to use


You can try the alpha.0 release in an existing plugin because it's been published to npm. 
```bash
cd theplugin
npm install @kinvolk/headlamp-plugin@0.6.0-alpha.0
npx @kinvolk/headlamp-plugin@0.6.0-alpha.0 upgrade --skip-package-updates
npm run storybook
```


To try out an example with a story see plugins/examples/pod-counter

```bash
git clone git@github.com:kinvolk/headlamp.git
cd headlamp
git fetch --all
git checkout plugin-source
cd plugins/examples/pod-counter
npm install @kinvolk/headlamp-plugin@0.6.0-alpha.0
npx @kinvolk/headlamp-plugin@0.6.0-alpha.0 upgrade --skip-package-updates
ls src/
npm run storybook
```

See plugins/examples/pod-counter/src/Message.stories.tsx and src/Message.tsx for story examples.


```bash
cd plugins/headlamp-plugin
npm run build
npm pack
cd ../examples/pod-counter
npm i ../../headlamp-plugin/*.tgz
npm run storybook
```

## Testing done

- [x] Used in the plugins repo https://github.com/headlamp-k8s/plugins/pull/5 inside a github workflow.
- [x] Can run the storybooks. `cd plugins/examples/pod-counter && npm install && npm run storybook`
- [x] Can run the tests (with jest and storyshot). `cd plugins/examples/pod-counter && npm install && npm run test`
- [x] The existing plugin commands still work (build, tsc, and lint).
- [x] for `npm run update-dependencies`, add frontend/package.json dependencies, goes into headlamp-plugin/package.json
- [x] for `npm run check-dependencies`, add frontend/package.json dependencies and it fails. Add to headlamp-plugin/package.json and it fails with return code 1. Which is what you want, because packages are not synced in these cases.
- [x] version changes of deps are handled `npm run update-dependencies`, add frontend/package.json dependencies, goes into headlamp-plugin/package.json
- [x] version changes of deps are handled `npm run check-dependencies`, add frontend/package.json dependencies and it fails. Add to headlamp-plugin/package.json and it fails with return code 1. Which is what you want, because packages are not synced in these cases.
- [x] `npm run check-dependencies` is run in CI to check dependencies are synced with frontend.
- [x] `npx @kinvolk/headlamp-plugin upgrade` works with existing plugin/examples plugins
- [x] importing CRD inside a component works in the plugin and in stories. (`console.log(CRD)` shows the object)

joaquimrocha commented on plugins/headlamp-plugin/config/plugins-tsconfig.json:
```diff
@@ -16,12 +16,12 @@
       "@kinvolk/headlamp-plugin/lib/CommonComponents/*": [
-        "node_modules/@kinvolk/headlamp-plugin/types/components/common/*/index.d.ts",
-        "node_modules/@kinvolk/headlamp-plugin/types/components/common/*.d.ts"
+        "node_modules/@kinvolk/headlamp-plugin/lib/components/common/*/index.d.ts",
+        "node_modules/@kinvolk/headlamp-plugin/lib/components/common/*.d.ts"
```
I guess now you can point to *.ts rather than .d.ts?

joaquimrocha commented on plugins/headlamp-plugin/package.json:
```diff
@@ -4,10 +4,13 @@
+    "build": "cd ../../frontend && tsc --build ./tsconfig.plugins.json && cd ../plugins/headlamp-plugin && npx shx cp -r ../../frontend/src/assets lib/ && npx shx cp -r ../../frontend/src/resources lib/ && npx shx cp -r ../../frontend/src/i18n/locales lib/i18n"
   },
   "dependencies": {
+    "@apidevtools/json-schema-ref-parser": "^9.0.9",
```
If this is mostly a copy of the dependencies we have in the frontend/ folder, should we try to have this file be generated from a template one? So we only have to care about dependencies in one place.

illume replied:
There's a script now to sync dependencies. There's also a ci check to make sure they're synced.

joaquimrocha reviewed (commented):
About shipping the .js . Is this really necessary?
I wonder if we could generate the js on the fly as a pre step somehow. The difference to other/regular npm packages is that `headlamp-plugin` can be used to bootstrap/manage the plugin, so this step may be accomplishable.

joaquimrocha commented on plugins/headlamp-plugin/template/package.json:
```diff
@@ -29,6 +29,9 @@
   "devDependencies": {
-    "@kinvolk/headlamp-plugin": "^$${headlamp-plugin-version}"
+    "@kinvolk/headlamp-plugin": "^$${headlamp-plugin-version}",
+    "react": "17.0.2",
+    "react-dom": "17.0.2",
+    "@storybook/addon-storyshots": "6.4.13"
```
We've been trying to avoid shipping dependencies beyond headlamp-plugin, so we'd not have to take care of keeping them under control. But if there's no workaround for it, I guess we have to use them..

joaquimrocha commented on plugins/headlamp-plugin/test-headlamp-plugin.js:
```diff
@@ -103,6 +103,7 @@ function cleanup() {
     .forEach(file => fs.rmSync(file));
 
   const foldersToRemove = [path.join('.plugins', PACKAGE_NAME), PACKAGE_NAME];
+  console.log('Temp foldersToRemove', foldersToRemove);
```
Leftover from debug, or is this intentional?

joaquimrocha reviewed (commented):
What a massive effort. Overall the changes look good though they're too many to wrap my head around in a short time.
Left a few comments + nits of things I noticed. Some of them maybe apply only to older commits in the PR, so maybe it's worth also simplifying things by squashing where adequate.

illume commented:
Hey hey. This could use another review.

All the commands used in testing now work on folders of plugins as well. A github workflow for plugins is in the docs and also used in the plugins repo https://github.com/headlamp-k8s/plugins/pull/5

🧹🧽 I've removed all the intermediate alpha release commits, and squashed a few of the commits too.

illume commented on plugins/examples/pod-counter/package.json:
```diff
@@ -31,6 +34,6 @@
   },
   "devDependencies": {
-    "@kinvolk/headlamp-plugin": "^0.5.6"
+    "@kinvolk/headlamp-plugin": "^0.6.0-alpha.15"
```
this should be updated to latest

illume commented on plugins/examples/pod-counter/package.json:
```diff
@@ -9,7 +9,10 @@
     "format": "headlamp-plugin format",
     "lint": "headlamp-plugin lint",
-    "tsc": "tsc --noEmit"
+    "tsc": "tsc --noEmit",
```
this needs updating too

illume commented on plugins/headlamp-plugin/README.md:
```diff
@@ -11,18 +11,20 @@ This package is published to the npm package index separately from Headlamp.
+headlamp-plugin upgrade <package>   Upgrade the plugin to latest headlamp-plugin.
+                                    Audits, formats, lints and type checks.
+                                    <package> defaults to current working
+                                    directory. Can also be a folder of packages.
```
Need updated usage docs from headlamp-plugin command. Format/lint/etc are missing and not updated for multiple packages.

illume commented on plugins/headlamp-plugin/bin/headlamp-plugin.js:
```diff
@@ -449,25 +617,44 @@ function upgrade(packageFolder, skipPackageUpdates) {
+    let configChanged = false;
+
+    /**
+     * Replaces nested config keys with ones in the template.
```
Expand on this/clarify this comment.

illume commented on plugins/headlamp-plugin/dependencies-sync.js:
```diff
@@ -0,0 +1,210 @@
+// @ts-check
+'use strict';
+
+// Aids maintenance syncing packages between frontend/ and headlamp-plugin
```
Expand on this, elaborate... this comment.

ashu8912 commented:
Since we change how `import CRD, { KubeCRD, makeCustomResourceClass } from '@kinvolk/headlamp-plugin/lib/k8s/crd'; working.` works, This makes it a breaking change for plugins that use such an import style for `CRD`, one way we could fix this would be to update all such imports in users `src` directory when they update there plugin to this version

illume commented:
An example of using react testing library in the plugin should be done in a separate PR (make issue).

illume commented:
TODO: Bug: when updating frontend/package.json dependency to a new version `npm run update-dependencies` and `npm run check-dependencies` do not work.

illume commented:
The version changes in frontend/package.json are synced to headlamp-plugin/package.json now too.

ashu8912 reviewed (approved):
Tested, Nice work @illume 🎉
